nixos/modules/installer/sd-card/sd-image-aarch64.nix

```diff
@@ -14,9 +14,10 @@
   boot.consoleLogLevel = lib.mkDefault 7;
 
   # The serial ports listed here are:
+  # - ttyS2: for RockPro64 and Rock64 (1500000 baud to match the bootloader)
   # - ttyS0: for Tegra (Jetson TX1)
   # - ttyAMA0: for QEMU's -machine virt
-  boot.kernelParams = ["console=ttyS0,115200n8" "console=ttyAMA0,115200n8" "console=tty0"];
+  boot.kernelParams = [ "console=ttyS2,1500000n8" "console=ttyS0,115200n8" "console=ttyAMA0,115200n8" "console=tty0" ];
```

Comment:
This might not work as expected *universally*.

E.g. if there is still a `tty0` or `ttyAMA0` or `ttyS0`, the console will only bind to the "rightmost valid one".

This is because we don't have console multiplexing. There's WIP work for it (https://github.com/NixOS/nixpkgs/pull/123039) but it's stuck on the *input* side of things. (Stuck is not the right term... nothing has been attempted yet, and I haven't gotten around to looking at the problem since.)

There's actually a solution for this: there should be *no console= parameters*, which will resort to use the `/chosen/stdout` (IIRC) FDT node to select the appropriate console... Except that this is not being provided correctly because of... reasons... (not a NixOS issue).

This is uh... a bit of a mess.

I *don't know* what should be done here. There isn't a clean fool-proof existing solution.

Reply:
I only got console/boot/login output on the serial port when I put the `ttyS2` entry first. That sounds like it'd be related, but in this case it binds to the left-most one. Is that the behaviour you'd currently expect?

Sounds like for now I should just leave it alone and make sure the Rock[Pro]64 wiki page mentions how to edit `extlinux.conf` to add the console parameter, and says to make sure to have it in your NixOS configuration.